components/portal_players/wikis/smash/portal_players_custom.lua

```diff
@@ -0,0 +1,139 @@
+
+---Entry Point. Builds the player portal
+---@param frame Frame
+---@return Html
+function CustomPortalPlayers.run(frame)
+	local args = Arguments.getArgs(frame)
+
+	local portalPlayers = PortalPlayers(args)
+
+	portalPlayers.header = CustomPortalPlayers.header
+	portalPlayers.row = CustomPortalPlayers.row
+
+	return portalPlayers:create()
+end
+
+---Builds the header for the table
+---@param flag string
+---@param playerType string?
+---@return Html
+function CustomPortalPlayers:header(flag, showLocalizedName, playerType)
+	local teamText = String.isNotEmpty(playerType) and ' Team' or ' Team and Role'
+
+	local header = mw.html.create('tr')
+		:tag('th')
+			:attr('colspan', 5)
+			:css('padding-left', '1em')
+			:wikitext(flag .. ' ' .. (playerType or NON_PLAYER_HEADER))
+			:done()
+
+	local subHeader = mw.html.create('tr')
+		:tag('th'):css('width', '175px'):wikitext(' ID'):done()
+		:tag('th'):css('width', '175px'):wikitext(' Real Name'):done()
+		:tag('th'):css('width', '52px'):wikitext(' Main'):done()
+		:tag('th'):css('width', '250px'):wikitext(teamText):done()
+		:tag('th'):css('width', '120px'):wikitext(' Links'):done()
+
+	return mw.html.create()
+		:node(header)
+		:node(subHeader)
+end
+
+---Builds a table row
+---@param player table
+---@param isPlayer boolean
+---@return Html
+function CustomPortalPlayers:row(player, showLocalizedName, isPlayer)
+	local row = mw.html.create('tr')
+		:addClass(BACKGROUND_CLASSES[(player.status or ''):lower()])
+
+	row:tag('td'):wikitext(' '):node(OpponentDisplay.BlockOpponent{opponent = PortalPlayers.toOpponent(player)})
+	row:tag('td')
+		:wikitext(' ' .. player.name)
+		:wikitext(self.showLocalizedName and (' (' .. player.localizedname .. ')') or nil)
+
+	row:tag('td'):node(CustomPortalPlayers._getMainCharIcons(player))
+
+	local role = not isPlayer and mw.language.getContentLanguage():ucfirst((player.extradata or {}).role or '')
+	local teamText = mw.ext.TeamTemplate.teamexists(player.team) and Team.team(nil, player.team) or ''
+	if role and String.isEmpty(teamText) then
+		teamText = role
+	elseif String.isNotEmpty(role) then
+		teamText = teamText .. ' (' .. role .. ')'
+	end
+	row:tag('td'):wikitext(' ' .. teamText)
+
+	local links = Array.extractValues(Table.map(player.links or {}, function(key, link)
+		return key, ' [' .. link .. ' ' .. Links.makeIcon(Links.removeAppendedNumber(key), 25) .. ']'
+	end) or {}, Table.iter.spairs)
+
+	row:tag('td')
+		:addClass('plainlinks')
+		:css('line-height', '25px')
+		:css('padding', '1px 2px 1px 2px')
+		:css('max-width', '112px')
+		:wikitext(table.concat(links))
+
+	return row
+end
+
+---Builds the main cahracter display
+---@param player table
+---@return Html?
+function CustomPortalPlayers._getMainCharIcons(player)
+	if String.isEmpty(player.extradata.maingame) then
+		return
+	end
+
+	local activeGame = player.extradata.maingame
+
+	if String.isEmpty(player.extradata['main' .. activeGame]) then
+		mw.log(player.pagename, 'main' .. activeGame)
```

Comment:
```suggestion
```
?
or better error msg

Reply:
it is a (functional) copy of the current implementation on several wikis
would we want to let their player portals error in case of invalid maingame stuff?

Reply:
Doesn't need to `error()`, but I think the logged message can be improved
also CC @spazer

Reply:
maybe
```suggestion
		mw.log('Missing "main' .. activeGame .. '" in extradata on player page ' .. player.pagename)
```

Reply:
Sounds good to me